types/express-serve-static-core/index.d.ts

```diff
@@ -463,7 +463,12 @@ export interface Request<P extends Params = ParamsDictionary, ResBody = any, Req
 
     params: P;
 
-    query: any;
+    query: {
+        [key: string]:
+            // The value type here is a "poor man's `unknown`". When these types support TypeScript
+            // 3.0+, we can replace this with `unknown`.
+            {} | null | undefined
+    };
```

Comment:
Alternatively we could use this type:

```ts
type ParsedQs = { [key: string]: string | string[] | ParsedQs };
```

That's the return type of `qs.parse` which is the default query parser: https://expressjs.com/en/api.html#app.settings.table

If a custom query parser is used, users can add assert the type:

```typescript
type MyQuery = {
    q: number;
};

app.get('/', req => {
    const query = req.query as MyQuery

    query.q; // $ExpectType number
})
```